[MSE] Assertion in WebKit::RemoteMediaSourceProxy::addSourceBuffer

jyavenard opened the pull request:
#### 7bd7bbbe097cfa549fbdaa60357c2d3a4c9cb6ba
<pre>
[MSE] Assertion in WebKit::RemoteMediaSourceProxy::addSourceBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=291932">https://bugs.webkit.org/show_bug.cgi?id=291932</a>
<a href="https://rdar.apple.com/149290374">rdar://149290374</a>

Reviewed by Youenn Fablet.

The completion handler given to RemoteMediaSourceProxy::addSourceBuffer
wasn&apos;t always called.
Ensure the completion handler is called when the RemoteMediaSourceProxy is
in a bad state, including when RemoteMediaSourceProxy::setPrivateAndOpen wasn&apos;t called.

Add crash test.
* LayoutTests/ipc/invalid-addSourceBuffer-to-GPU-process-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-addSourceBuffer-to-GPU-process-crash.html: Added.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::addSourceBuffer):

Canonical link: <a href="https://commits.webkit.org/294043@main">https://commits.webkit.org/294043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83edb0357b3de3d0de0f5a278f2b2a1f6f206ff1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100660 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10611 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105797 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51248 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102701 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28786 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76649 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33689 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90920 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57004 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15638 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8927 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50624 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85550 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108152 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27778 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20398 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85604 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28141 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87121 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85145 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21664 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29847 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7573 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21766 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27713 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32966 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27524 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30842 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29082 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
```diff
@@ -123,11 +123,14 @@ void RemoteMediaSourceProxy::failedToCreateRenderer(RendererType)
 {
     RefPtr connection = connectionToWebProcess();
-    if (!m_remoteMediaPlayerProxy || !connection)
+    RefPtr msp = mediaSourcePrivate();
```
```suggestion
    RefPtr mediaSourcePrivate = this->mediaSourcePrivate();
```

jyavenard committed:
[MSE] Assertion in WebKit::RemoteMediaSourceProxy::addSourceBuffer
https://bugs.webkit.org/show_bug.cgi?id=291932
rdar://149290374

Reviewed by Youenn Fablet.

The completion handler given to RemoteMediaSourceProxy::addSourceBuffer
wasn't always called.
Ensure the completion handler is called when the RemoteMediaSourceProxy is
in a bad state, including when RemoteMediaSourceProxy::setPrivateAndOpen wasn't called.

Add crash test.
* LayoutTests/ipc/invalid-addSourceBuffer-to-GPU-process-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-addSourceBuffer-to-GPU-process-crash.html: Added.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::addSourceBuffer):

Canonical link: https://commits.webkit.org/294043@main
